File: ToDoModel/Dimension Libraries/Project/ToDoProjectInfoEntry.vb
Imports UtilitiesVB

<Serializable()>
Public Class ToDoProjectInfoEntry
    Implements IEquatable(Of IToDoProjectInfoEntry),
        IComparable, IComparable(Of IToDoProjectInfoEntry), IToDoProjectInfoEntry

    Public Property ProjectName As String Implements IToDoProjectInfoEntry.ProjectName
    Public Property ProjectID As String Implements IToDoProjectInfoEntry.ProjectID
    Public Property ProgramName As String Implements IToDoProjectInfoEntry.ProgramName

    Public Sub New(ByVal ProjName As String, ProjID As String, ProgName As String)
        ProjectName = ProjName
        ProjectID = ProjID
        ProgramName = ProgName
    End Sub

    Public Overloads Function Equals(other As IToDoProjectInfoEntry) _
        As Boolean Implements IEquatable(Of IToDoProjectInfoEntry).Equals,
        IToDoProjectInfoEntry.Equals

        Return other IsNot Nothing AndAlso ProjectName.Equals(other.ProjectName)
    End Function

    Public Overrides Function Equals(obj As Object) As Boolean Implements IToDoProjectInfoEntry.Equals
        If obj Is Nothing Then Return False

        Dim other As ToDoProjectInfoEntry = TryCast(obj, ToDoProjectInfoEntry)
        Return other IsNot Nothing AndAlso Equals(other)
    End Function

    Public Function CompareTo(other As IToDoProjectInfoEntry) As Integer Implements IComparable(Of IToDoProjectInfoEntry).CompareTo, IToDoProjectInfoEntry.CompareTo
        If other Is Nothing Then
            Return 1
        Else
            Dim x As Integer = String.CompareOrdinal(ProjectID, other.ProjectID)
            If x = 0 Then
                If ProjectID.Length < other.ProjectID.Length Then
                    x = -1
                ElseIf ProjectID.Length > other.ProjectID.Length Then
                    x = 1
                End If
            End If
            Return x
            'Return Me.ProjectID.CompareTo(other.ProjectID)
        End If
    End Function

    Public Function CompareTo(obj As Object) As Integer Implements IComparable.CompareTo
        If obj Is Nothing Then Return 1
        Dim other As IToDoProjectInfoEntry = TryCast(obj, IToDoProjectInfoEntry)

        If other IsNot Nothing Then
            Return CompareTo(other)
        Else
            Throw New ArgumentException("Object cannot be cast to IToDoProjectInfoEntry")
        End If
    End Function

    Public Function ToCSV() As String Implements IToDoProjectInfoEntry.ToCSV
        Return ProjectID + "," + ProjectName + "," + ProgramName
    End Function

End Class

File: ToDoModel/Dimension Libraries/Project/ToDoProjectInfoEntry.cs
using System;
using UtilitiesVB;
using UtilitiesCS;
using System.Collections.Generic;

namespace ToDoModel
{

    [Serializable()]
    public class ToDoProjectInfoEntry : IEquatable<IToDoProjectInfoEntry>, IComparable, IComparable<IToDoProjectInfoEntry>, IToDoProjectInfoEntry, IEquatable<ToDoProjectInfoEntry>
    {

        private string _projectName;
        private string _projectID;
        private string _programName;

        public string ProjectName { get => _projectName; set => _projectName = value; }
        public string ProjectID { get => _projectID; set => _projectID = value; }
        public string ProgramName { get => _programName; set => _programName = value; }

        public ToDoProjectInfoEntry(string ProjName, string ProjID, string ProgName)
        {
            ProjectName = ProjName;
            ProjectID = ProjID;
            ProgramName = ProgName;
        }

        public int CompareTo(IToDoProjectInfoEntry other)
        {
            if (other is null) { return 1; }
            else if (ProjectID is null) { return -1; }
            else
            {
                int x = string.CompareOrdinal(ProjectID, other.ProjectID);
                if (x == 0)
                {
                    if (ProjectID.Length < other.ProjectID.Length)
                    {
                        x = -1;
                    }
                    else if (ProjectID.Length > other.ProjectID.Length)
                    {
                        x = 1;
                    }
                }
                return x;
                // Return Me.ProjectID.CompareTo(Other.ProjectID)
            }
        }

        public int CompareTo(object obj)
        {
            if (obj is null)
                return 1;
            IToDoProjectInfoEntry other = obj as IToDoProjectInfoEntry;

            if (other is not null)
            {
                return CompareTo(other);
            }
            else
            {
                throw new ArgumentException("Object cannot be cast to IToDoProjectInfoEntry");
            }
        }

        public string ToCSV()
        {
            return ProjectID + "," + ProjectName + "," + ProgramName;
        }

        public override bool Equals(object obj)
        {
            return Equals(obj as IToDoProjectInfoEntry);
        }

        public bool Equals(ToDoProjectInfoEntry other)
        {
            return other is not null &&
                   ProjectName == other.ProjectName &&
                   ProjectID == other.ProjectID &&
                   ProgramName == other.ProgramName;
        }

        public bool Equals(IToDoProjectInfoEntry other)
        {
            return other is not null &&
                   ProjectName == other.ProjectName &&
                   ProjectID == other.ProjectID &&
                   ProgramName == other.ProgramName;
        }

        public override int GetHashCode()
        {
            int hashCode = 682028280;
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(ProjectName);
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(ProjectID);
            hashCode = hashCode * -1521134295 + EqualityComparer<string>.Default.GetHashCode(ProgramName);
            return hashCode;
        }
    
        public bool IsAnyNull() 
        { return (_projectName is null)||(_projectID is null)||(_programName is null); }
    }
}